Dockerfile

```diff
@@ -1,64 +1,79 @@
-
-    # Upgrade the package index and install security upgrades
-RUN --mount=type=cache,target=/var/lib/apt/lists \
-    --mount=type=cache,target=/var/cache,sharing=locked \
-    apt-get update \
-    && apt-get upgrade --assume-yes \
-    && apt-get install --assume-yes --no-install-recommends python3-pip
-
-# Install dependencies
-FROM python-base as poetry
-
-WORKDIR /app
-# that contains poetry=<version> to install a specific version of Poetry.
-COPY requirements.txt ./
-RUN --mount=type=cache,target=/root/.cache python3.11 -m pip install --disable-pip-version-check --requirement=requirements.txt
-
-# Do the conversion
-COPY poetry.lock pyproject.toml ./
-RUN poetry export --output=requirements.txt
-
-FROM python-base as development
-
-RUN --mount=type=cache,target=/root/.cache --mount=type=bind,from=poetry,source=/app,target=/poetry python3.11 -m pip install --disable-pip-version-check --no-deps --requirement=/poetry/requirements.txt
+    PIP_DEFAULT_TIMEOUT=100 \
+    # Disable poetry interaction
+    POETRY_NO_INTERACTION=1 \
+    POETRY_VIRTUALENVS_IN_PROJECT=1 \
+    POETRY_VIRTUALENVS_CREATE=1 \
+    POETRY_CACHE_DIR=/tmp/poetry_cache
+
+RUN pip install poetry==${POETRY_VERSION}
+
+# Install the app. Just copy the files needed to install the dependencies
+COPY pyproject.toml poetry.lock README.md ./
+
+# Poetry cache is used to avoid installing the dependencies every time the code changes, we will keep this folder in development environment and remove it in production
+# --no-root, poetry will install only the dependencies avoiding to install the project itself, we will install the project in the final layer
+# --without dev to avoid installing dev dependencies, we do not need test and linters in production environment
+# --with dev to install dev dependencies, we need test and linters in development environment
+# --mount, mount a folder for plugins with poetry cache, this will speed up the process of building the image
+RUN if [ {${DEV}} ]; then \
+      echo "Installing dev dependencies"; \
+      poetry install --no-root --with dev \
+    else \
+      echo "Skipping dev dependencies"; \
+      poetry install --no-root --without dev && rm -rf ${POETRY_CACHE_DIR}; \
+   fi
+
+# Set up our final runtime layer
+FROM python:3.11-slim-bookworm as runtime
+
+ENV VIRTUAL_ENV=/app/.venv \
+    PATH="/app/.venv/bin:$PATH"
+
+COPY --from=base ${VIRTUAL_ENV} ${VIRTUAL_ENV}
+
+# Create our users here in the last layer or else it will be lost in the previous discarded layers
+# Create a system group named "app_user" with the -r flag
+RUN groupadd -r app_user \
+    # Create a system user named "app_user" with the -r flag
+    && useradd -r -g app_user app_user
+
+# Switch to the non-root user "user"
+USER app_user
 
 WORKDIR /app
 
-COPY . .
+COPY . /app
 
-CMD ["python3"]
+# Activate the virtual environment
+CMD . /app/venv/bin/activate
 
-USER appuser
+CMD ["tail", "-f", "/dev/null"]
```

Comment:
When it comes to executing a command to launch the container, it's better to use only one CMD directive. If you use more than one, only the last one will be called and used for the container. It is also against best practices to have more than one. Remember that the CMD directive can be over-written by any other commands passed via the cli when running the container.

Reply:
thanks K'Ron, I will change the last CMD by an ENTRYPOINT

Reply:
You will want to have `ENTRYPOINT` before `CMD`